File: api/app/schemas/revive.py
from datetime import datetime
from decimal import Decimal
from enum import Enum
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class PaymentStatus(str, Enum):
    PENDING = "PENDING"
    FAILED = "FAILED"
    SUCCESS = "SUCCESS"
    ABANDONED = "ABANDONED"
    UNRESOLVED = "UNRESOLVED"


class RootCause(str, Enum):
    PAYMENT_TIMEOUT = "payment_timeout"
    BANK_DECLINE = "bank_decline"
    AUTHENTICATION_FAILURE = "authentication_failure"
    INSUFFICIENT_FUNDS = "insufficient_funds"
    PAYMENT_METHOD_ISSUE = "payment_method_issue"
    CUSTOMER_ABANDONMENT = "customer_abandonment"
    TECHNICAL_FAILURE = "technical_failure"
    REPEATED_PAYMENT_FAILURE = "repeated_payment_failure"
    UNKNOWN = "unknown"


class RecommendedAction(str, Enum):
    CONTROLLED_RETRY = "controlled_retry"
    RECOVERY_REMINDER = "recovery_reminder"
    RETRY_AUTHENTICATION = "retry_authentication"
    ESCALATE_HUMAN = "escalate_human"
    STOP_RECOVERY = "stop_recovery"
    NO_ACTION = "no_action"


class CustomerOut(BaseModel):
    id: str
    name: str
    email: str | None = None
    phone: str | None = None
    status: str
    created_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class TransactionCreate(BaseModel):
    transaction_id: str = Field(min_length=2, max_length=120)
    customer_id: str = Field(min_length=2, max_length=64)
    order_id: str = Field(min_length=2, max_length=120)
    amount: Decimal = Field(ge=Decimal("0.00"))
    currency: str = Field(default="INR", min_length=3, max_length=8)
    payment_method: str = Field(min_length=2, max_length=80)
    status: PaymentStatus = PaymentStatus.PENDING
    failure_reason: str | None = None
    gateway_response: str | None = None
    retry_count: int = Field(default=0, ge=0)
    customer_response: str | None = None
    customer_name: str = Field(default="Guest Customer", max_length=255)
    customer_email: str | None = Field(default=None, max_length=255)
    customer_phone: str | None = None


class TransactionUpdate(BaseModel):
    status: PaymentStatus | None = None
    failure_reason: str | None = None
    gateway_response: str | None = None
    retry_count: int | None = Field(default=None, ge=0)
    customer_response: str | None = None
    recovery_status: str | None = None
    escalation_status: str | None = None


class PaymentAttemptOut(BaseModel):
    id: str
    transaction_id: str
    attempt_number: int
    status: str
    gateway_response: str | None
    created_at: datetime | None

    model_config = ConfigDict(from_attributes=True)


class RecoveryActionOut(BaseModel):
    id: str
    recovery_case_id: str
    action_type: str
    action_reason: str
    policy_result: dict[str, Any]
    execution_result: dict[str, Any] | None
    status: str
    idempotency_key: str | None
    created_at: datetime | None

    model_config = ConfigDict(from_attributes=True)


class RecoveryCaseOut(BaseModel):
    id: str
    transaction_id: str
    risk_amount: Decimal
    root_cause: str | None
    confidence: float | None
    evidence: list[str]
    recommended_action: str | None
    action_status: str
    recovery_status: str
    recovered_amount: Decimal
    policy_result: dict[str, Any] | None
    detection_timestamp: datetime | None
    success_timestamp: datetime | None
    created_at: datetime | None
    updated_at: datetime | None

    model_config = ConfigDict(from_attributes=True)


class TransactionOut(BaseModel):
    id: str
    transaction_id: str
    customer_id: str
    order_id: str
    amount: Decimal
    currency: str
    payment_method: str
    status: str
    failure_reason: str | None
    gateway_response: str | None
    retry_count: int
    customer_response: str | None
    recovery_status: str
    recovered_amount: Decimal
    escalation_status: str
    created_at: datetime | None
    updated_at: datetime | None
    customer: CustomerOut | None = None

    model_config = ConfigDict(from_attributes=True)


class TransactionDetailOut(TransactionOut):
    payment_attempts: list[PaymentAttemptOut] = []
    recovery_cases: list[RecoveryCaseOut] = []


class DiagnosisResult(BaseModel):
    transaction_id: str
    revenue_at_risk: Decimal
    root_cause: RootCause
    confidence: float = Field(ge=0.0, le=1.0)
    evidence: list[str] = Field(min_length=1, max_length=5)
    reason: str = Field(min_length=1, max_length=500)
    requires_human_review: bool


class DecisionResult(BaseModel):
    transaction_id: str
    revenue_at_risk: Decimal
    root_cause: RootCause
    confidence: float = Field(ge=0.0, le=1.0)
    recommended_action: RecommendedAction
    reason: str = Field(min_length=1, max_length=500)
    evidence: list[str] = Field(min_length=1, max_length=5)
    requires_human_review: bool


class PolicyValidationOut(BaseModel):
    result: Literal["APPROVED", "BLOCKED", "ESCALATE"]
    allowed: bool
    reasons: list[str]
    max_automatic_retries: int
    max_recovery_messages: int
    current_retry_count: int
    current_message_count: int


class RecoveryStartRequest(BaseModel):
    idempotency_key: str | None = Field(default=None, max_length=255)
    force_payment_result: Literal["SUCCESS", "FAILED"] | None = None


class PaymentRetryRequest(BaseModel):
    idempotency_key: str | None = Field(default=None, max_length=255)
    force_result: Literal["SUCCESS", "FAILED"] | None = None


class AuditLogOut(BaseModel):
    id: str
    transaction_id: str | None
    recovery_case_id: str | None
    event_type: str
    event_message: str
    actor: str
    metadata: dict[str, Any]
    created_at: datetime | None


class EscalationOut(BaseModel):
    id: str
    transaction_id: str
    recovery_case_id: str | None
    reason: str
    priority: str
    status: str
    ai_recommendation: str | None
    action_history: list[dict[str, Any]]
    created_at: datetime | None
    resolved_at: datetime | None

    model_config = ConfigDict(from_attributes=True)


class DashboardSummary(BaseModel):
    total_transactions: int
    failed_transactions: int
    revenue_at_risk: Decimal
    recovery_attempts: int
    successful_recoveries: int
    revenue_recovered: Decimal
    recovery_rate: float
    unresolved_cases: int
    escalated_cases: int
    failure_rate: float
    revenue_recovery_rate: float
    average_recovery_latency_seconds: float | None

File: frontend/types/revive.ts
export type PaymentStatus = "PENDING" | "FAILED" | "SUCCESS" | "ABANDONED" | "UNRESOLVED";

export type RecoveryStatus =
  | "OPEN"
  | "IN_PROGRESS"
  | "RECOVERED"
  | "FAILED"
  | "ESCALATED"
  | "STOPPED"
  | "UNRESOLVED";

export type ActionStatus =
  | "PENDING"
  | "POLICY_APPROVED"
  | "POLICY_BLOCKED"
  | "EXECUTED"
  | "FAILED";

export type EscalationStatus = "NONE" | "OPEN" | "IN_REVIEW" | "RESOLVED";

export type EscalationPriority = "LOW" | "MEDIUM" | "HIGH" | "CRITICAL";

export type RootCauseType =
  | "payment_timeout"
  | "bank_decline"
  | "authentication_failure"
  | "insufficient_funds"
  | "payment_method_issue"
  | "customer_abandonment"
  | "technical_failure"
  | "repeated_payment_failure"
  | "unknown";

export type RecommendedActionType =
  | "controlled_retry"
  | "recovery_reminder"
  | "retry_authentication"
  | "escalate_human"
  | "stop_recovery"
  | "no_action";

export interface Customer {
  id: string;
  name: string;
  email: string | null;
  phone: string | null;
  status: string;
  created_at?: string | null;
}

export interface PaymentAttempt {
  id: string;
  transaction_id: string;
  attempt_number: number;
  status: PaymentStatus;
  gateway_response?: string | null;
  created_at?: string | null;
}

export interface PolicyResult {
  result?: "APPROVED" | "BLOCKED" | "ESCALATE" | "REJECT";
  allowed: boolean;
  policy?: "ALLOW" | "REJECT" | "ESCALATE";
  action?: string;
  reasons?: string[];
  reason?: string;
  max_automatic_retries?: number;
  max_recovery_messages?: number;
  current_retry_count?: number;
  current_message_count?: number;
  customer_opt_out?: boolean;
}

export interface RecoveryAction {
  id: string;
  recovery_case_id: string;
  action_type: RecommendedActionType | string;
  action_reason: string;
  policy_result: PolicyResult;
  execution_result?: Record<string, any> | null;
  status: ActionStatus;
  idempotency_key?: string | null;
  created_at?: string | null;
}

export interface RecoveryCase {
  id: string;
  transaction_id: string;
  risk_amount: number;
  root_cause?: RootCauseType | string | null;
  confidence?: number | null;
  evidence: string[];
  recommended_action?: RecommendedActionType | string | null;
  action_status: ActionStatus;
  recovery_status: RecoveryStatus;
  recovered_amount: number;
  policy_result?: PolicyResult | null;
  detection_timestamp?: string | null;
  success_timestamp?: string | null;
  created_at?: string | null;
  updated_at?: string | null;
  recovery_actions?: RecoveryAction[];
}

export interface Transaction {
  id: string;
  transaction_id: string;
  customer_id: string;
  order_id: string;
  amount: number;
  currency: string;
  payment_method: string;
  status: PaymentStatus;
  failure_reason?: string | null;
  gateway_response?: string | null;
  retry_count: number;
  customer_response?: string | null;
  recovery_status: RecoveryStatus;
  recovered_amount: number;
  escalation_status: EscalationStatus;
  created_at?: string | null;
  updated_at?: string | null;
  customer?: Customer | null;
  payment_attempts?: PaymentAttempt[];
  recovery_cases?: RecoveryCase[];
}

export interface PaginatedTransactions {
  data: Transaction[];
  pagination: {
    limit: number;
    offset: number;
    returned: number;
    next_offset: number | null;
  };
}

export interface DashboardSummary {
  total_transactions: number;
  failed_transactions: number;
  revenue_at_risk: number;
  total_risk_detected: number;
  recovery_attempts: number;
  successful_recoveries: number;
  revenue_recovered: number;
  recovery_rate: number;
  unresolved_cases: number;
  escalated_cases: number;
  failure_rate: number;
  revenue_recovery_rate: number;
  average_recovery_latency_seconds?: number | null;
  generated_at?: string;
}

export interface RecoveryMetrics {
  summary: DashboardSummary;
  case_status_counts: Record<string, number>;
}

export interface DiagnosisResult {
  transaction_id: string;
  revenue_at_risk?: number;
  root_cause: RootCauseType | string;
  confidence: number;
  evidence: string[];
  reason: string;
  requires_human_review: boolean;
}

export interface DecisionResult {
  transaction_id: string;
  root_cause: string;
  confidence: number;
  decision: RecommendedActionType | string;
  policy: string;
  allowed: boolean;
  reason: string;
  requires_human_review: boolean;
  escalation_id?: string | null;
}

export interface RecoveryStartResponse {
  transaction_id: string;
  decision?: DecisionResult | Record<string, any>;
  policy?: PolicyResult;
  recovery_case?: RecoveryCase | null;
  action_id?: string;
  execution_result?: Record<string, any>;
  escalation_id?: string;
  payment_status?: string;
  recovery_status?: string;
  recovered_amount?: number;
  transaction?: Transaction | any;
}

export interface EscalationCase {
  id: string;
  transaction_id: string;
  recovery_case_id?: string | null;
  reason: string;
  priority: EscalationPriority;
  status: "OPEN" | "IN_REVIEW" | "RESOLVED";
  ai_recommendation?: string | null;
  action_history: Array<{
    action?: string;
    reason?: string;
    timestamp?: string;
    [key: string]: any;
  }>;
  created_at?: string | null;
  resolved_at?: string | null;
  transaction?: Transaction | null;
}
